### PR TITLE
More ERC20 token Incentive in gauge

### PR DESCRIPTION
**Pull request by ludete**

### CODE REVIEW

File: contracts/Voter.sol [hunk 1]
```diff
@@ -386,4 +414,23 @@ contract Voter is IVoter, ERC2771Context, ReentrancyGuard {
...
     }
   }
+
+  function depositLP(address _pool, uint256 _amount) external hasValidGauge(_pool) nonReentrant {
+    if (_amount == 0) revert ZeroAmount();
+    address _sender = msg.sender;
+    address _gauge = gauges[_pool];
+    IERC20(_pool).safeTransferFrom(_sender, address(this), _amount);
+    IERC20(_pool).approve(_gauge, _amount);
+    IGauge(_gauge).deposit(_amount, _sender);
```
Comment (CoderZhi) [inline contracts/Voter.sol:424]: users are able to deposit to a gauge directly, as a result, the incentive won't be updated. Meanwhile, gaugeToIncentives is the only thing needed. Thus, could we move incentive of a gauge into the gauge, such that when there is a deposit to the gauge, we can update its corresponding incentive.

File: contracts/rewards/CustomReward.sol [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import {Reward} from "./Reward.sol";
+
+/// @title Base voting reward contract for distribution of rewards by token id
+///        on a weekly basis
+abstract contract CustomReward is Reward {
+  constructor(address _forwarder, address _voter, address[] memory _rewards) Reward(_forwarder, _voter) {
+    uint256 _length = _rewards.length;
+    for (uint256 i; i < _length; i++) {
+      if (_rewards[i] != address(0)) {
```
[the remark this answers is not in the document]
Comment (ludete) [reply contracts/rewards/CustomReward.sol:12]: has ignore it when _rewards[i]==0

File: contracts/Voter.sol [hunk 3]
```diff
@@ -360,6 +379,15 @@ contract Voter is IVoter, ERC2771Context, ReentrancyGuard {
...
   }
 
+  /// @inheritdoc IVoter
+  function claimIncentive(address[] memory _incentives, address[][] memory _tokens) external {
+    address _sender = msg.sender;
+    uint256 _length = _incentives.length;
+    for (uint256 i = 0; i < _length; i++) {
+      IReward(_incentives[i]).getReward(_sender, _tokens[i]);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ludete) [reply contracts/Voter.sol:389]: del

File: contracts/interfaces/IReward.sol [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  function numCheckpoints(address user) external view returns (uint256);
+
+  /// @notice The total number of checkpoints
+  function supplyNumCheckpoints() external view returns (uint256);
+
+  /// @notice Deposit an amount into the rewards contract to earn future rewards associated to a veNFT
+  /// @dev Internal notation used as only callable internally by `authorized`.
+  /// @param amount   Amount deposited for the veNFT
+  /// @param user  Which deposit LP
+  function _deposit(uint256 amount, address user) external;
```
[the remark this answers is not in the document]
Comment (ludete) [reply contracts/interfaces/IReward.sol:71]: the external func is only called by voter or later by gauge when change.

File: contracts/Voter.sol (hunk 1 above)
Comment (ludete) [reply contracts/Voter.sol:424]: Gauge and incentive are associated with each other and callback is called in gauge when deposit.

File: contracts/rewards/CustomReward.sol [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+pragma solidity ^0.8.0;
+
+import {Reward} from "./Reward.sol";
+
+/// @title Base voting reward contract for distribution of rewards by token id
+///        on a weekly basis
+abstract contract CustomReward is Reward {
+  constructor(address _forwarder, address _voter, address[] memory _rewards) Reward(_forwarder, _voter) {
+    uint256 _length = _rewards.length;
+    for (uint256 i; i < _length; i++) {
```
[the remark this answers is not in the document]
Comment (ludete) [reply contracts/rewards/CustomReward.sol:11]: No restrictions are needed from a business perspective

File: contracts/rewards/Reward.sol [hunk 6]
```diff
@@ -46,10 +46,20 @@ abstract contract Reward is IReward, ERC2771Context, ReentrancyGuard {
...
+  /// @inheritdoc IReward
+  address public gauge;
+
   constructor(address _forwarder, address _voter) ERC2771Context(_forwarder) {
     voter = _voter;
   }
 
+  function setGauge(address _gauge) external {
+    require(gauge == address (0), "gauge has been set");
+    if (msg.sender != authorized) revert NotAuthorized();
```
Comment (CoderZhi) [inline contracts/rewards/Reward.sol:58]: authorized is not initialized

File: contracts/rewards/Reward.sol [hunk 7]
```diff
@@ -153,6 +163,10 @@ abstract contract Reward is IReward, ERC2771Context, ReentrancyGuard {
     return rewards.length;
   }
 
+  function rewardTokens() external view returns (address[] memory) {
+    return rewards;
```
Comment (CoderZhi) [inline contracts/rewards/Reward.sol:167]: rewards may be too long to be returned

File: contracts/rewards/Reward.sol [hunk 8]
```diff
@@ -246,4 +257,13 @@ abstract contract Reward is IReward, ERC2771Context, ReentrancyGuard {
 
     emit NotifyReward(sender, token, epochStart, amount);
   }
+
+  /// @dev add new reward token to incentive.
+  function addRewardToken(address token) external {
```
Comment (CoderZhi) [inline contracts/rewards/Reward.sol:262]: who can add reward token?

File: contracts/rewards/Reward.sol (hunk 6 above)
Comment (ludete) [reply contracts/rewards/Reward.sol:58]: initialized in CustomReward when construct.

File: contracts/rewards/Incentive.sol [hunk 9]
```diff
@@ -15,13 +15,6 @@ contract Incentive is CustomReward {
   /// @inheritdoc CustomReward
   function notifyRewardAmount(address token, uint256 amount) external override nonReentrant {
     address sender = _msgSender();
-
-    if (!isReward[token]) {
-      if (!IVoter(voter).isWhitelistedToken(token)) revert NotWhitelisted();
-      isReward[token] = true;
-      rewards.push(token);
-    }
```
[the remark this answers is not in the document]
Comment (ludete) [reply contracts/rewards/Incentive.sol]: The latest commit was not committed in local, which fix the issue.

File: contracts/rewards/Reward.sol (hunk 8 above)
Comment (ludete) [reply contracts/rewards/Reward.sol:262]: everyone, which the new add token must be WhitelistedToken in Voter.

Comment (CoderZhi) [reply contracts/rewards/Reward.sol:262]: in that case,  what's the point of adding the reward token address to Reward.sol?

**Pull request closed without merging**